// taapi/src/dataProvider/TaapiCache.js
const axios = require("axios");

class ExchangeDataTaapiCache
{
    constructor(secret, endpoint) {
        this.secret = secret;
        this.endpoint = endpoint;
    }

    async getCandles(provider, source, symbol, interval, limit) {
       
        let candles = [];

        if(this.endpoint) {

            await axios.get(`${this.endpoint}/candles`, { params: {
                secret: this.secret,
                provider: provider,
                source: source,
                symbol: symbol,
                interval: interval,
                limit: limit,
            }}).then( response => {

                // Axios wraps the response data in a data object and so does the taapicache API
                candles = response.data.data;
                
            })
            .catch(function (error) {
                console.error(error);
            });

        } else {
            console.error("No endpoint defined for data provider 'taapicache'!");
        }
        
        return candles;
    }
}

module.exports = ExchangeDataTaapiCache;
