// taapi/src/Errors.js
/**
 * Define error types that may be returned by this API
 */
const errorTypes = {
    OK: {
      message: "200: Request OK!",
      statusCode: 200
    },
    BAD_REQUEST: {
      message: "400: Bad request. Check your request parameters and try again!",
      statusCode: 400
    },
    INVALID_EXCHANGE: {
      message: "400: Bad request. This exchange is not support with this method. Use the client method to calculate TA from other exchanges than Binance!",
      statusCode: 400
    },
    UNAUTHORIZED: {
      message: "401: You are not authenticated to query this endpoint!",
      statusCode: 401
    },
    FORBIDDEN: {
      message: "403: Permission denied for this endpoint!",
      statusCode: 403
    },
    NOT_FOUND: {
      message: "404: The endpoint you are calling does not exist!",
      statusCode: 404
    },
    TOO_MANY_REQUESTS: {
      message: "429: You have exceeded your request limit (rate-limit)!",
      statusCode: 429
    },
    INTERNAL_SERVER_ERROR: {
      message: "500: An internal server error has occurred!",
      statusCode: 500
    }
  };
  
  module.exports = {ErrorTypes: errorTypes};
